File: packages/moshe-sdk-python/src/moshe/_telemetry.py
from __future__ import annotations

import uuid
from dataclasses import dataclass

from ._interfaces import TelemetrySink
from ._types import Decision, FeedbackVerdict, TelemetryEvent, TelemetryEventType, clone_dataclass


class MemoryTelemetrySink(TelemetrySink):
    name = "memory"

    def __init__(self) -> None:
        self._events: list[TelemetryEvent] = []

    async def emit(self, event: TelemetryEvent) -> None:
        self._events.append(clone_dataclass(event))

    def get_events(self) -> list[TelemetryEvent]:
        return [clone_dataclass(event) for event in self._events]

    def get_events_by_type(self, event_type: TelemetryEventType) -> list[TelemetryEvent]:
        return [clone_dataclass(event) for event in self._events if event.event_type == event_type]

    def get_events_for_action(self, action_id: str) -> list[TelemetryEvent]:
        return [clone_dataclass(event) for event in self._events if event.action_id == action_id]

    def get_decision_event(self, action_id: str) -> TelemetryEvent | None:
        matches = [
            event
            for event in self._events
            if event.action_id == action_id and event.event_type == "DECISION_MADE"
        ]
        return clone_dataclass(matches[-1]) if matches else None

    def clear(self) -> None:
        self._events.clear()


@dataclass(frozen=True)
class FeedbackSubmission:
    action_id: str
    session_id: str
    verdict: FeedbackVerdict
    expected_decision: Decision | None = None
    note: str | None = None


class FeedbackEmitter:
    def __init__(self, sinks: list[TelemetrySink]) -> None:
        self._sinks = sinks

    async def submit(self, submission: FeedbackSubmission) -> None:
        event = TelemetryEvent(
            event_id=str(uuid.uuid4()),
            event_type="FEEDBACK",
            action_id=submission.action_id,
            session_id=submission.session_id,
            debug={
                "verdict": submission.verdict,
                **({"expected_decision": submission.expected_decision} if submission.expected_decision else {}),
                **({"note": submission.note} if submission.note is not None else {}),
            },
        )
        for sink in self._sinks:
            await sink.emit(event)

File: packages/moshe-core/src/telemetry.ts
import { randomUUID } from 'node:crypto';

import type { TelemetryEvent, TelemetryEventType } from '@moshesdk/spec';

import type { TelemetrySink } from './interfaces.js';

export class MemoryTelemetrySink implements TelemetrySink {
  public readonly name = 'memory';
  private readonly events: TelemetryEvent[] = [];

  public async emit(event: TelemetryEvent): Promise<void> {
    this.events.push(event);
  }

  public getEvents(): readonly TelemetryEvent[] {
    return this.events;
  }

  public getEventsByType(eventType: TelemetryEventType): readonly TelemetryEvent[] {
    return this.events.filter((event) => event.eventType === eventType);
  }

  public getEventsForAction(actionId: string): readonly TelemetryEvent[] {
    return this.events.filter((event) => event.actionId === actionId);
  }

  public getDecisionEvent(actionId: string): TelemetryEvent | undefined {
    return this.events
      .filter((event) => event.actionId === actionId && event.eventType === 'DECISION_MADE')
      .at(-1);
  }

  public clear(): void {
    this.events.length = 0;
  }
}

export class ScrubbingTelemetrySink implements TelemetrySink {
  public readonly name: string;

  public constructor(private readonly inner: TelemetrySink) {
    this.name = `scrubbing(${inner.name})`;
  }

  public async emit(event: TelemetryEvent): Promise<void> {
    const { debug: _debug, ...scrubbed } = event;
    await this.inner.emit(scrubbed as TelemetryEvent);
  }
}

export interface FeedbackSubmission {
  actionId: string;
  sessionId: string;
  verdict: 'FALSE_POSITIVE' | 'FALSE_NEGATIVE' | 'CORRECT';
  expectedDecision?: 'ALLOW' | 'BLOCK' | 'REVIEW';
  note?: string;
}

export class FeedbackEmitter {
  public constructor(private readonly sinks: TelemetrySink[]) {}

  public async submit(feedback: FeedbackSubmission): Promise<void> {
    const event: TelemetryEvent = {
      eventId: randomUUID(),
      eventType: 'FEEDBACK',
      actionId: feedback.actionId,
      sessionId: feedback.sessionId,
      debug: {
        verdict: feedback.verdict,
        ...(feedback.expectedDecision !== undefined ? { expectedDecision: feedback.expectedDecision } : {}),
        ...(feedback.note !== undefined ? { note: feedback.note } : {})
      }
    };

    for (const sink of this.sinks) {
      await sink.emit(event);
    }
  }
}
